src/appleseed-max-impl/appleseedenvmap/appleseedenvmap.cpp

```diff
@@ -0,0 +1,588 @@
+
+    BaseClone(this, mnew, remap);
+    return (RefTargetHandle)mnew;
+}
+
+int AppleseedEnvMap::NumSubTexmaps()
+{
+    return TexmapCount;
+}
+
+Texmap* AppleseedEnvMap::GetSubTexmap(int i)
+{
+    return (i == 0) ? m_turbidity_map : nullptr;
+}
+
+void AppleseedEnvMap::SetSubTexmap(int i, Texmap* texmap)
+{
+    if (i > 0)
+      return;
+    ReplaceReference(i + 1, texmap);
+
+    const auto texmap_id = static_cast<TexmapId>(i);
+    const auto param_id = g_texmap_id_to_param_id[texmap_id];
+    m_pblock->SetValue(param_id, 0, texmap);
+    g_block_desc.InvalidateUI( param_id );
+    m_map_validity.SetEmpty();
+}
+
+int AppleseedEnvMap::MapSlotType(int i)
+{
+    return MAPSLOT_TEXTURE;
+}
+
+TSTR AppleseedEnvMap::GetSubTexmapSlotName(int i)
+{
+    const auto texmap_id = static_cast<TexmapId>(i);
+    return g_texmap_slot_names[texmap_id];
+}
+
+void AppleseedEnvMap::Update(TimeValue t, Interval& valid)
+{
+    if (!m_params_validity.InInterval(t))
+    {
+        m_params_validity.SetInfinite();
+
+        m_pblock->GetValue(ParamIdSunTheta, t, m_sun_theta, m_params_validity);
+        m_pblock->GetValue(ParamIdSunPhi, t, m_sun_phi, m_params_validity);
+
+        m_pblock->GetValue(ParamIdTurbidity, t, m_turbidity, m_params_validity);
+        m_pblock->GetValue(ParamIdTurbidityMap, t, m_turbidity_map, m_params_validity);
+        m_pblock->GetValue(ParamIdTurbidityMapOn, t, m_turbidity_map_on, m_params_validity);
+
+        m_pblock->GetValue(ParamIdTurbMultiplier, t, m_turb_multiplier, m_params_validity);
+        m_pblock->GetValue(ParamIdLuminMultiplier, t, m_lumin_multiplier, m_params_validity);
+        m_pblock->GetValue(ParamIdLuminGamma, t, m_lumin_gamma, m_params_validity);
+        m_pblock->GetValue(ParamIdSatMultiplier, t, m_sat_multiplier, m_params_validity);
+        m_pblock->GetValue(ParamIdHorizonShift, t, m_horizon_shift, m_params_validity);
+        m_pblock->GetValue(ParamIdGroundAlbedo, t, m_ground_albedo, m_params_validity);
+
+        NotifyDependents(FOREVER, PART_ALL, REFMSG_CHANGE);
+    }
+
+    if (!m_map_validity.InInterval(t))
+	{
+		m_map_validity.SetInfinite();
+	    if (m_turbidity_map)
+	        m_turbidity_map->Update(t,m_map_validity);
+	}
+
+    valid &= m_map_validity;
+    valid &= m_params_validity;
+}
+
+void AppleseedEnvMap::Reset()
+{
+    m_params_validity.SetEmpty();
+    m_map_validity.SetEmpty();
+}
+
+Interval AppleseedEnvMap::Validity(TimeValue t)
+{
+    Interval valid = FOREVER;
+    Update(t, valid);
+    return valid;
+}
+
+ParamDlg* AppleseedEnvMap::CreateParamDlg(HWND hwMtlEdit, IMtlParams* imp)
+{
+    IAutoMParamDlg* masterDlg = g_appleseed_appleseedenvmap_classdesc.CreateParamDlgs(hwMtlEdit, imp, this);
+    g_block_desc.SetUserDlgProc(new EnvMapParamMapDlgProc());
+
+    return masterDlg;
+}
+
+const USHORT ChunkFileFormatVersion                 = 0x0001;
+const USHORT ChunkMtlBase                           = 0x1000;
+const USHORT FileFormatVersion = 0x0001;
```

Comment:
Can I use chunk id-s from datachunk.h header? Documantation says "The ID passed need only be unique within the plug-ins data itself." That means unique across this Save method I suppose? Not across all the plugins?